Added spice blooms

Mailaender opened the pull request:
https://github.com/OpenRA/OpenRA/pull/8627 with unrelated stuff moved into it's own commit and removed the `Health.Resurect` hack.

Closes #4462.

Unit158 committed:
document ThrowsShrapnel

Unit158 committed:
Moved RandomWalk to OpenRA.Traits.Util

Mailaender committed:
add an empty sprite

Mailaender committed:
add spice blooms

Unit158 commented:
With the delayed action, you don't need to use the counter when it is dead.

On 2015-08-08, at 9:53 AM, Matthias Mailänder <notifications@github.com> wrote:

> #8627 with unrelated stuff moved into it's own commit and removed the Health.Resurect hack.
> 
> You can view, comment on, or merge this pull request online at:
> 
>   https://github.com/OpenRA/OpenRA/pull/8987
> 
> Commit Summary
> 
> Moved RandomWalk to OpenRA.Traits.Util
> document ThrowsShrapnel
> add an empty sprite
> add spiceblooms
> File Changes
> 
> M OpenRA.Game/Traits/Util.cs (15)
> M OpenRA.Mods.Common/Traits/SeedsResource.cs (17)
> M OpenRA.Mods.Common/Traits/ThrowsShrapnel.cs (12)
> M OpenRA.Mods.D2k/OpenRA.Mods.D2k.csproj (1)
> A OpenRA.Mods.D2k/Traits/SpiceBloom.cs (169)
> D mods/d2k/bits/spicebloom.shp     (0)
> M mods/d2k/rules/arrakis.yaml (25)
> M mods/d2k/rules/defaults.yaml (6)
> M mods/d2k/sequences/misc.yaml     (17)
> M mods/d2k/weapons.yaml (9)
> Patch Links:
> 
> https://github.com/OpenRA/OpenRA/pull/8987.patch
> https://github.com/OpenRA/OpenRA/pull/8987.diff
> —
> Reply to this email directly or view it on GitHub.
>

pchote commented on OpenRA.Mods.D2k/Traits/SpiceBloom.cs:
```diff
@@ -0,0 +1,169 @@
+			this.info = info;
+			self = init.Self;
+
+			resType = self.World.WorldActor.TraitsImplementing<ResourceType>().First(t => t.Info.Name == info.ResourceType);
```
Might want to think about using *OrDefault and then manually throwing a more understandable exception if there are no matches (same for the line below).

pchote commented on OpenRA.Mods.D2k/Traits/SpiceBloom.cs:
```diff
@@ -0,0 +1,169 @@
+			}
+			else
+			{
+				counter++;
```
I thought we were going to do this with aggregated damage so that attacking the spawned would accelerate the spawning instead of having two separate mechanisms that don't interact.

Unit158 replied:
This code and the code you mentioned below are obsolete as I mentioned above.

pchote commented on OpenRA.Mods.D2k/Traits/SpiceBloom.cs:
```diff
@@ -0,0 +1,169 @@
+					anim.Animation.Play(info.GrowthSequences[index]);
+				}
+			}
+			else
```
Is this branch actually used?

Unit158 replied:
Nope.

Mailaender commented:
Cleaned this up further. Fixed the crushing and adapted the balancing so that early (forced) explosion don't give you as much as pieces as fully grown once so you can only speed up, but not generate more by manual targeting.

GraionDilach commented on OpenRA.Mods.D2k/Traits/SpiceBloom.cs:
```diff
@@ -0,0 +1,169 @@
+				});
+			}
+
+			self.World.AddFrameEndTask(w => w.Add(new DelayedAction(respawnTicks, () =>
```
It'd be nice if this would be optional, then this could be used for a crop logic - using minelayers for seeding.

Mailaender replied:
This isn't really a general purpose trait designed to be re-usable. Better write your own. I also don't quite get what you have in mind here to be honest.

GraionDilach replied:
Alright.

Unit158 replied:
You can actually already do that. Explodes, SelfHeals (negative heal) and a spreads resource warhead or whatever it's called.

GraionDilach replied:
Good point on SelfHeal, I totally forgot that could be used for this.

Unit158 commented on OpenRA.Mods.D2k/Traits/SpiceBloom.cs:
```diff
@@ -0,0 +1,168 @@
+			})));
+		}
+
+		public void WarnCrush(Actor crusher) { }
```
This should be removed and replaced by the generic Crushable trait, no?

Mailaender replied:
That didn't work out. As discussed in IRC that is quite infantry specific with `WarnCrush` overhead etc.

pchote replied:
The `Crushable` trait works fine for walls.

Mailaender replied:
Actually it doesn't. I made a ticket for it. https://github.com/OpenRA/OpenRA/issues/9011

pchote replied:
Fixed by #9014.

Mailaender replied:
Rebased and fixed the Devastator crush class while I was at it.

Mailaender committed:
make Devastators crush walls

Unit158 commented on mods/d2k/weapons.yaml:
```diff
@@ -717,8 +717,15 @@ Shrapnel:
+		High: true
+		Angle: 91, 264
+		Trail: large_trail
+		Image: large_trail
```
This may or may not want to be changed to the "null" sprite. I think It looks a little bit better like this, however.

abcdefg30 commented:
Needs a rebase.

Mailaender commented:
Rebased.

pchote commented on OpenRA.Mods.D2k/Traits/SpiceBloom.cs:
```diff
@@ -0,0 +1,154 @@
+		{
+			if (!self.IsDead)
+			{
+				ticks++;
```
I'm still not a big fan of having a separate tick counter for this.  Like I mentioned earlier, this means that there is no interaction between damage and the timer - they are both all or nothing.  If this instead decremented the health then shooting the spice bloom to half health would halve the spawn time, which I believe is more intuitive.

Unit158 replied:
Ok, but that isn't the original behaviour. This replicates the original behaviour (nearly) to a T, minus some heuristics that were included in the original to only spawn the spiceblooms when there was a certain amount of spice needed.

On 2015-08-09, at 4:49 PM, Paul Chote <notifications@github.com> wrote:

> In OpenRA.Mods.D2k/Traits/SpiceBloom.cs:
> 
> > +			resType = self.World.WorldActor.TraitsImplementing<ResourceType>().First(t => t.Info.Name == info.ResourceType);
> > +
> > +			var render = self.Trait<RenderSprites>();
> > +			anim = new AnimationWithOffset(new Animation(init.Self.World, render.GetImage(self)), null, () => self.IsDead);
> > +			render.Add(anim);
> > +
> > +			respawnTicks = self.World.SharedRandom.Next(info.RespawnDelay[0], info.RespawnDelay[1]);
> > +			growTicks = self.World.SharedRandom.Next(info.GrowthDelay[0], info.GrowthDelay[1]);
> > +			anim.Animation.Play(info.GrowthSequences[0]);
> > +		}
> > +
> > +		public void Tick(Actor self)
> > +		{
> > +			if (!self.IsDead)
> > +			{
> > +				ticks++;
> I'm still not a big fan of having a separate tick counter for this. Like I mentioned earlier, this means that there is no interaction between damage and the timer - they are both all or nothing. If this instead decremented the health then shooting the spice bloom to half health would halve the spawn time, which I believe is more intuitive.
> 
> ―
> Reply to this email directly or view it on GitHub.
>

pchote replied:
Ok, fair enough.

abcdefg30 commented:
This causes weird glitches:
![spicebloomglitch](https://cloud.githubusercontent.com/assets/7704140/9365598/b22000ca-46b2-11e5-964e-afe066706be0.png)

penev92 commented:
Regarding that glitch - @OmegaBolt mentioned on IRC yesterday that he's planning a PR to change the terrain type of sand decorations so that spice doesn't grow on them and replace them. Maybe this is related.

Mailaender commented:
Yes, those are just terrain definitions errors like https://github.com/OpenRA/OpenRA/issues/4243.

Mailaender commented:
Updated according to the review comments.

pchote commented:
Adding [Changes Requested] for the MCV/Harv fixes, and needs a rebase.

Mailaender commented:
Rebased and fixed the crushable regressions.

abcdefg30 commented:
Looks good to me. :+1: / :white_check_mark:

reaperrr commented:
Needs rebase.

Mailaender commented:
Rebased.

reaperrr commented:
Sorry, needs another rebase.

Mailaender commented:
Rebased.

penev92 commented on OpenRA.Mods.D2k/Traits/SpiceBloom.cs:
```diff
@@ -0,0 +1,153 @@
+
+		public void Tick(Actor self)
+		{
+			if (!self.IsDead)
```
The trait shouldn't tick if the actor is dead, so this should not be necessary.

penev92 commented on OpenRA.Mods.D2k/Traits/SpiceBloom.cs:
```diff
@@ -0,0 +1,153 @@
+
+		public void Killed(Actor self, AttackInfo e)
+		{
+			var weapon = self.World.Map.Rules.Weapons[info.Weapon.ToLowerInvariant()];
```
You can get rid of the local variable as it is only used once anyway.

penev92 commented on mods/d2k/sequences/misc.yaml:
```diff
@@ -334,19 +334,20 @@ crate:
 		ZOffset: -511
 		Offset: -16,-16
 
-# TODO: keep the redundant spicebloom.shp for now for the awful WinForms maps editor
```
Perhaps with the removal of this comment (and the long-gone WinForms map editor) you should remove that SHP?

Mailaender replied:
I already did that.

penev92 commented:
Also the actor that crushes the spicebloom is unharmed. Not sure if it should be instakilled or dealt a specific amount of damage, but there should be damaging involved.

Aside from that this looks OK. :+0.5: for the moment.

Mailaender commented:
Updated.

penev92 commented:
Sorry this still doesn't harm the actor that crushes the spicebloom.

obrakmann commented:
Yep. The problem seems to be that the SpreadDamageWarhead is fired at the cells where the spice is spread to, not at the location where the spice bloom was.

penev92 commented:
Added this to the Next release milestone because it will compliment the D2k rebalancing nicely and is almost ready IMO.

GraionDilach commented:
SpiceExplosion's warhead needs AffectsParent for the crusher to be harmed.

pchote commented:
Needs a rebase.

Mailaender committed:
let the crusher take damage on spice explosions

Mailaender commented:
Updated.

reaperrr commented:
There are still quite a few differences compared to the original, but I think this is now good enough to build upon. :+1:

pchote commented:
What are the differences?

reaperrr commented:
[changelog](https://github.com/OpenRA/OpenRA/wiki/Changelog-%28bleed%29/_compare/0976c2833a4a6662620cff9005d16d12e78348e7...2253cc95eb70037f241edf9e0d1113a079d52ad7)

> What are the differences?

Spice debris were flying slower and on much lower angles (and more of them, I think) and spice blooms were only growing when there was no spice on that tile. They were growing slower, too.

I don't mind filing a follow-up myself, though.

reaperrr commented:
@Mailaender Actually, I didn't notice this before due to the large variance in Angle, but all fired spice debris target the same cell, something isn't right with the 'target' randomization.

pchote commented on OpenRA.Mods.D2k/Traits/SpiceBloom.cs:
```diff
@@ -0,0 +1,148 @@
+				if (cell == null)
+					cell = cells.Take(info.Range).Random(self.World.SharedRandom);
+
+				args.PassiveTarget = self.World.Map.CenterOfCell(cell.Value);
```
This is why.  This gets reset each time through the loop, so all the projectiles will be spawned with the last value that is assigned.

Mailaender replied:
Actually this is not the problem. This has to be set for each projectile. The randomization a few lines before just doesn't work, because `cell` is always null.
